flax/linen/module.py

```diff
@@ -1526,6 +1527,43 @@ def init(self,
         **kwargs)
     return v_out
 
+  @traceback_util.api_boundary
+  def lazy_init(self,
+           rngs: Union[PRNGKey, RNGSequences],
+           *args,
+           method: Optional[Callable[..., Any]] = None,
+           mutable: CollectionFilter = DenyList('intermediates'),
+           **kwargs) -> FrozenVariableDict:
+    """Initializes a module without computing on an actual input.
+
+    lazy_init will initialize the variables without doing unnecessary compute.
+    The input data should be passed as a ``jax.ShapeDtypeStruct`` which specifies
```

Comment:
Can we accept any type that has `.shape` + `.dtype` and convert it to `ShapeDtypeStruct` under the hood? It would be nice if users could just switch from `init` to `lazy_init` without further changes.

Reply:
Previously we did something magic like that but sometimes you need a real value instead a ShapeDtypeStruct for this to work

Reply:
done